cfme/utils/single.py

```diff
@@ -0,0 +1,24 @@
+def single(iterable):
```

Comment:
I'm confused by this function. It doesn't seem to work if there is more than a single item in the iterable. Example, 
```python,
single([1,2,3])
Traceback (most recent call last):
  File "/home/jdupuy/cfme_tests/cfme_venv/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2878, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-20-2bda16e9a15f>", line 1, in <module>
    single([1,2,3])
  File "/home/jdupuy/cfme_tests/integration_tests/cfme/utils/single.py", line 16, in single
    raise MoreThanOne('There are more than only one item in {!r}'.format(iterable))
MoreThanOne: There are more than only one item in [1, 2, 3]
```

Is this behavior desired? Is so, why not just use the index of 0?

Reply:
Yes, it is desired. The point of this function is to raise an exception when you get more items somewhere where you would expect one item only, Like, for example, from function `get_master`. You would expect one item back, but in replicated environments you get more.

Reply:
I'd agree with John here that what you've described in your comment is not obvious expected behavior. There's no mention of that behavior, just 'checks'.

I've got a few issues with this implementation and use, and will try to break them down into specific points, as they tend to intertwine.

This is being used in places where there's no additional handling of these new exception types that are raised, where before no exception was raised when there were multiple items in iterable, and would have raised an IndexError. That IndexError was at least predictable in some way - `ItemMissing` is not.

These exceptions should likely be defined in a more central location, like cfme/exceptions.py, if they're to be used.

This implementation seems far too over complicated to me, for what its doing. Python has a syntax for exactly this behavior, but it raises far more expected built in exception types.

Using the tuple unpacking syntax, a `ValueError` is raised in the case that the iterable is empty, or has multiple values.  This should be used when expecting a single value in an iterable ... or asserting its length is `1` (to be very explicit).

If I want to use this function for the same behavior, I now have to import the whole module, or the function and the two exception defintions separately. Versus no imports, and just having to inspect the exception message if I want to base some following logic on too many vs empty. Again, that logic could be based on the length of the iterable if I don't want to use a try/except structure.

```
In [7]: val, = [1]

In [8]: val
Out[8]: 1

In [9]: val, = []
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-e3b3d792319b> in <module>()
----> 1 val, = []

ValueError: need more than 0 values to unpack

In [10]: val, = [1,2]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-9bfe3baaea3f> in <module>()
----> 1 val, = [1,2]

ValueError: too many values to unpack


```

Reply:
Nice! it seems like the tuple unpacking does what I needed.